docs/design/features/objc-interop.md

```diff
@@ -0,0 +1,160 @@
+
+An Objective-C Block variable defined as:
+
+```objective-c
+int(^blk)(int) = ^(int a) { return a * 2; };
+```
+
+Can be queried, following the ABI, for its invoke function pointer and dispatched in C as:
+
+```C
+void* fptr = extract_using_abi(blk);
+int b = ((int(*)(id,int))fptr)(blk, a);
+```
+
+The lifetime of Blocks initially relies upon semantics similar to stack clean-up in C++. This makes lifetime management more complicated in .NET. A key take away here is that the runtime will always initially create a Block that is, according to Objective-C semantics, stack allocated. The contract for Blocks and how reference counting works means that if the Objective-C caller requires the Block longer than the current calling scope the Block should be copied (i.e. [`Block_copy`](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Blocks/Articles/bxUsing.html#//apple_ref/doc/uid/TP40007502-CH5-SW2)) and when the current managed calling scope is left the created Block released.
+
+**Method dispatch**
+
+The Objective-C [message dispatch mechanism](#message_disp) will require special handling if either exception propagation or variadic variable signatures are desired. If neither exceptions nor variadic arguments are a concern then acquiring the native function pointer to the appropriate message function would be sufficient for internal uses - C# function pointers make this very easy. However, as officially described [here](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors), users are permitted to declare their own P/Invoke signatures and perform manual message dispatch - this complicates things.
+
+Xamarin presently defines a set of hand-written assembly for the messaging functions and they handle execptions and most variadic argument cases. Then, utilizing Mono's [`mono_dllmap_insert`](http://docs.go-mono.com/?link=root:/embed) API, maps these entries to override the official Objective-C runtime APIs when loaded through a P/Invoke. This results in users only needing to know the official Objective-C runtime APIs but getting all the benefits of the hand-written assembly.
+
+There are several possible approaches to address this going forward:
+
+* Support a `mono_dllmap_insert` type API to permit overloading.
+* Hardcode these special cases in the runtime with the assembly code.
+* Hardcode these special cases in the runtime but permit the "overrides" to be provided through a managed API.
+
+The last option seems to have the most flexibility and follows the "pay-for-play" principle since some users may not care about exception handling or variadic argument support.
```

Comment:
How does `PINVOKE_OVERRIDE` fit into this vs the approach in https://github.com/dotnet/runtime/pull/47721? There is quite an overlap.

`PINVOKE_OVERRIDE` could easily be used by the native code to implement the same logic as `mono_dllmap_insert` used. It requires support from the host though.

I am not convinced that a managed interface is very helpful in this case. The `objc_*` methods themselves still have to be implemented in the native code due to the nature of the argument dispatch. Who would register the callbacks and at what time would that happen? If the answer is that it's done once at startup and driven by the host then I think the approach is unnecessarily complicated and duplicates the functionality that is already available through other means.

Reply:
The `PINVOKE_OVERRIDE` logic was leveraged to implement this feature. The overlap here is related to implementation detail only since the `PINVOKE_OVERRIDE` feature isn't publicly exposed. There is no desire to presently expose that and thus we permit a narrow override for this specific scenario. I am assuming that `PINVOKE_OVERRIDE` is referring to what is used to handle single file work.

>I am not convinced that a managed interface is very helpful in this case.

CoreCLR is not exposing functionality through native means. Using `UnmanagedCallersOnly` and function pointers exposing most functionality to native can be done in a bespoke manner based on a consumer's needs.

>I am not convinced that a managed interface is very helpful in this case. The objc_* methods themselves still have to be implemented in the native code due to the nature of the argument dispatch.

They don't actually need to be implemented in native at all depending on what one wants to do with them. We have various tests that implement them in managed code using `UnmanagedCallersOnly`. This isn't to say we can handle all functionality but most is possible at least with what we've prototyped - perhaps there is an example?

>Who would register the callbacks and at what time would that happen? If the answer is that it's done once at startup and driven by the host then I think the approach is unnecessarily complicated and duplicates the functionality that is already available through other means.

Yep, this must be done relatively early.

Reply:
`PINVOKE_OVERRIDE` is for unmanaged hosts. It cannot be really replaced by managed code for all cases. The host cannot inject the managed code soon enough for it to kick in.

Our original design goal was to make Objective-C interop work without special host. For example, to allow using libraries that use nice Objective-C interop in regular console applications, with regular host that is not aware that Objective-C interop exists. It is how we have the designed the Com/WinRT interop for .NET and we are happy with it.

It is not clear whether we will be able to maintain this goal due to resource constrains. If we end up giving up on this goal and make the Objective-C interop work with just the Xamarin/Apple app host, I agree that this method is not really needed.

Reply:
> They don't actually need to be implemented in native at all depending on what one wants to do with them.

I have seen the code in the tests of the feature/objc-interop branch but I had the feeling that it covers only narrow scenario. I was thinking specifically about the implementation in Xamarin that does translation of Objective-C exceptions. I may need to read more into it though.

> Our original design goal was to make Objective-C interop work without special host.

I think this somewhat slipped off my mind and it's a valid point. On the other hand it feels like `SetMessageSendCallback` specifically is problematic if two different libraries tried to use it. Likely not a big deal in practice but I remember seeing the concern in some comment earlier.

Reply:
> I was thinking specifically about the implementation in Xamarin that does translation of Objective-C exceptions.

If one wants exceptions support, I agree that is likely not possible to implement in C#, but should not be the reason for forcing it be a hard requirement.

> On the other hand it feels like SetMessageSendCallback specifically is problematic if two different libraries tried to use it.

Yes that is true. This API is really a niche case for Xamarin at this point. The `ComWrappers` also has an issue with singleton usage to - the GC interaction model. It was a desire to make that more narrow but given Xamarin established guidance around usage of [P/Invoke with Selectors](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors#calling-objc_msgsend) it is hard to have a model that can work with multiple consumers.

Reply:
I guess my point was that if it's primarily for the Xamarin use-case then PINVOKE_OVERRIDE would work there just as good. It has the additional benefit of being slightly more flexible and guaranteeing the registration soon enough to ensure that all P/Invokes are consistently resolved (and thus it's easier to debug and reason about). Since there can only be one host it also takes care of the issue of two consumers trying to use the interface. Lastly, it's available on Mono and CoreCLR today so the implementation for both runtimes could be easily shared on the Xamarin side.

Obviously that was all under the assumption that one would have the control of the unmanaged host. If this is supposed to cover other scenarios like running under the default CoreCLR host (corehost, corerun or other) and possibly a managed implementation of the callbacks then I see the reason for the different approach.

I am still unsure what can you actually achieve in the managed code and how would you handle all the cases of parameter passing and forwarding that to the underlying implementation. The [unit test](https://github.com/elinor-fung/runtime/blob/272c91f0bb35cb56207724f4bde28383fb5430d4/src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ObjectiveC/MessageSendTests.cs) doesn't seem to cover that ground.